From a47c6c8bce778715a4589b4bba02da5922a20819 Mon Sep 17 00:00:00 2001
From: Yang Jihong <yangjihong1@huawei.com>
Date: Wed, 6 Jul 2022 18:05:27 +0800
Subject: [PATCH] livepatch/arm64: Add arch_klp_module_check_calltrace
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: a47c6c8bce778715a4589b4bba02da5922a20819
Modified-by-SEL: No


hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I5CJ7X

--------------------------------

Add arch_klp_module_check_calltrace to check whether stacks of all tasks
are within the code segment of module.

Signed-off-by: Yang Jihong <yangjihong1@huawei.com>
Reviewed-by: Xu Kuohai <xukuohai@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/include/asm/livepatch.h |  1 +
 arch/arm64/kernel/livepatch.c      | 23 +++++++++++++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/arch/arm64/include/asm/livepatch.h b/arch/arm64/include/asm/livepatch.h
index b87dc35c2b3f..bcb6c4081978 100644
--- a/arch/arm64/include/asm/livepatch.h
+++ b/arch/arm64/include/asm/livepatch.h
@@ -69,6 +69,7 @@ struct arch_klp_data {
 int arch_klp_add_breakpoint(struct arch_klp_data *arch_data, void *old_func);
 void arch_klp_remove_breakpoint(struct arch_klp_data *arch_data, void *old_func);
 long arch_klp_save_old_code(struct arch_klp_data *arch_data, void *old_func);
+int arch_klp_module_check_calltrace(void *data);
 
 #endif
 
diff --git a/arch/arm64/kernel/livepatch.c b/arch/arm64/kernel/livepatch.c
index 539d6534a220..cda56066d859 100644
--- a/arch/arm64/kernel/livepatch.c
+++ b/arch/arm64/kernel/livepatch.c
@@ -68,6 +68,7 @@ struct klp_func_list {
 struct walk_stackframe_args {
 	int enable;
 	struct klp_func_list *check_funcs;
+	struct module *mod;
 	int ret;
 };
 
@@ -323,6 +324,28 @@ int klp_check_calltrace(struct klp_patch *patch, int enable)
 	return ret;
 }
 
+static bool check_module_calltrace(void *data, unsigned long pc)
+{
+	struct walk_stackframe_args *args = data;
+
+	if (within_module_core(pc, args->mod)) {
+		pr_err("module %s is in use!\n", args->mod->name);
+		args->ret = -EBUSY;
+		return false;
+	}
+	return true;
+}
+
+int arch_klp_module_check_calltrace(void *data)
+{
+	struct walk_stackframe_args args = {
+		.mod = (struct module *)data,
+		.ret = 0
+	};
+
+	return do_check_calltrace(&args, check_module_calltrace);
+}
+
 int arch_klp_add_breakpoint(struct arch_klp_data *arch_data, void *old_func)
 {
 	u32 insn = BRK64_OPCODE_KLP;
-- 
2.33.0

